cpu/stm32f1/periph/gpio.c

```diff
@@ -68,38 +68,32 @@ static inline int _pin_num(gpio_t pin)
 }
 
 
-int gpio_init(gpio_t pin, gpio_dir_t dir, gpio_pp_t pullup)
+int gpio_init(gpio_t pin, gpio_mode_t mode)
 {
     GPIO_TypeDef *port = _port(pin);
     int pin_num = _pin_num(pin);
 
+    /* open-drain output with pull-up is not supported */
+    if (mode == GPIO_OD_PU) {
+        return -1;
+    }
 
     /* enable the clock for the selected port */
     RCC->APB2ENR |= (RCC_APB2ENR_IOPAEN << _port_num(pin));
-    /* clear configuration */
+
+    /* set pin mode */
     port->CR[pin_num >> 3] &= ~(0xf << ((pin_num & 0x7) * 4));
-    /* set new configuration */
-    if (dir == GPIO_DIR_OUT) {
-        if (pullup != GPIO_NOPULL) {
-            return -1;
-        }
-        /* set to output, push-pull, 50MHz */
-        port->CR[pin_num >> 3] |= (0x3 << ((pin_num & 0x7) * 4));
-        /* clear pin */
-        port->BRR = (1 << pin_num);
-    }
-    else {
-        /* configure pin to input, pull register according to the value of
-         * the pullup parameter */
-        port->CR[pin_num >> 3] |= ((pullup & 0xc)  << ((pin_num & 0x7) * 4));
-        port->ODR &= ~(1 << pin_num);
-        port->ODR |=  ((pullup & 0x1) << pin_num);
+    port->CR[pin_num >> 3] |=  (mode << ((pin_num & 0x7) * 4));
+    /* set initial state of output register */
+    port->BRR = (1 << pin_num);
+    if (mode == GPIO_IN_PU) {
```

Comment:
because

     +    GPIO_IN_PD = GPIO_MODE(0, 2),   /**< input with pull-down */
     +    GPIO_IN_PU = GPIO_MODE(0, 2),   /**< input with pull-up */

is the same value, this if statement is also true for PD :)